### ISSUE TITLE
PodTemplateSpec Does not allow configuring existing containers

### ISSUE THREAD

**Issue by kri5** [datastax/cass-operator#108]
I would have supposed that the containers and initcontainers already used by the operator could be configured/overriden by specifying some keys on the PodTemplateSpec, and thoses would be merge on top of the default values.

Use case would be to pass to the cassandra config init container some environment values interpolated from secrets, or some volumemounts, but it seems that it just append the keys to the others, so i end up with configuration with multiple "server-config-init" init containers or multiple "cassandra" containers.

Is this the expected behavior?

**Comment by bradfordcp** [comment]
Related #101 

**Short Answer:** Yes

**Long Answer:**

A collection of `initContainers` is by design. We prepend our `initContainer` to render out the basic configuration which then allows the user provided `PodTemplateSpec.initContainers` to modify any values that have been generated.

I don't expect to see many `cassandra` containers in the same Pod. Do you have a use case for this?  Right now, we are _currently_ of the mindset that changes to the `cassandra` container should be handled with fields at the top-level of the `CassandraDatacenter` CR. 

It appears as though you expect that if you provide an entry in `PodTemplateSpec.containers` with the same `name` that any values applied to that container would be merged with the container spec prepared by the operator. Is that accurate?

What changes are you considering making to the `cassandra` container?

**Comment by kri5** [comment]
Below comes an example of what i am trying to achieve

```
    podTemplateSpec:                                                                                                
      spec:                                                                                                         
        volumes:                                                                                                    
          - name: truststore                                                                                        
            secret:                                                                                                 
              secretName: cassandra-tls                                                                         
        initContainers:                                                                                             
          - name: server-config-init                                                                                
            env:                                                                                                    
              - name: TRUSTSTORE_PASSWORD                                                                           
                valueFrom:                                                                                          
                  secretKeyRef:                                                                                     
                    name: cassandra-truststore-key                                                              
                    key: key                                                                                        
              - name: KEYSTORE_PASSWORD                                                                             
                valueFrom:                                                                                          
                  secretKeyRef:                                                                                     
                    name: cassandra-keystore-key                                                              
                    key: key                                                                                        
        containers:                                                                                                 
          - name: cassandra                                                                                         
            volumeMounts:                                                                                           
              - mountPath: /secrets/tls                                                                             
                name: truststore                                                                                    
                                                                                                                    
        tolerations:                                                                                                
          - key: dedicated                                                                                          
            operator: Equal                                                                                         
            value: cassandra                                                                                        
            effect: NoSchedule
...
  config:                                                                                                         
      cassandra-yaml:                                                                                                                                         
        server_encryption_options:
          keystore: /secrets/tls/keystore.jks
          keystore_password: $KEYSTORE_PASSWORD                                                                     
          truststore: /secrets/tls/truststore.jks                                                                   
          truststore_password: $TRUSTSTORE_PASSWORD 
```

What i want to achieve it being able to pass values from secrets via env to the init-container dealing with the configuration, and mounting volumes into the cassandra container, and also specifing tolerations on all thoses containers.

**Comment by jimdickinson** [comment]
Can you do this with a second init container? It's guaranteed to run after the config-builder init container.

**Comment by kri5** [comment]
I could indeed have another init container that contains exactly the same things as the `server-config-init` one, and add my config there, but it defeats the point of DRY.

**Comment by kri5** [comment]
And i do not see how i could pass custom volume mounts to the `cassandra` container

**Comment by jimdickinson** [comment]
I'll try to sketch up a working example this weekend of how you get secrets from an added init container, but maybe we need to create another volume mount for this purpose. Are you opposed to using the existing `/config` in your init container?

**Comment by bradfordcp** [comment]
There's no need to repeat the same things as `server-config-init`. That container is still run if you specify your own initContainers. See https://github.com/datastax/cass-operator/blob/master/operator/pkg/reconciliation/constructor.go#L493 where we prepend `server-config-init` to anything passed from the template.

Right now we don't allow for adding additional volumes to the `cassandra` container, but as @jimdickinson mentioned there _is_ `/config` available to use. I could see an implementation where your `podTemplateSpec` utilizes an init container that mounts secrets and /config. From there it would copy the secrets over to /config and update the appropriate config file entries (using sed or maybe a YAML aware script)

Admittedly this is a little hack-ish, but we are actively looking at how users leverage podTemplateSpec to inform future changes to take some of these sharp edges out.

**Comment by eldondevat** [comment]
@kri5 We'd be interested in your specific TLS strategy here. The initial issue description here is very general, but we should be able to support TLS in a feature-rich and cassandra-oriented way, so having a TLS-specific discussion might be worthwhile.

**Comment by kri5** [comment]
Basically, my ideas was to use the existing cass-config-builder initcontainer and append some values to it (https://github.com/datastax/cass-config-builder).

I want to generate the configuration in order to have node-to-node encryption using a truststore, that is password protected. The key of the truststore is held in a kubernetes secrets. So i thought that i could use Kubernetes Downward API to mount the secret values in the environment of the config container in order to pass thoses values to the container (Not even sure this would have worked)

Afterward i need to mount the secrets containing the certificates inside the cassandra pod.

To sum up i need to be able to interpolate values from secrets in the configuration initContainer, and being able to mount secrets and/or volumes inside the cassandra container.

 That is what my example code is trying to achieve with the current state of the operator.

**Comment by kri5** [comment]
Any thoughts on this topic? Any non-TLS configuration that would need the same benefits? (Interpolated values from secrets through env-var, volume mounts in cassandra/init container)

**Comment by jimdickinson** [comment]
Are you planning on having the truststore be different per pod / Cassandra node?

**Comment by eldondevat** [comment]
@kri5 We're still discussing the use cases we have internally. I think the reason for the previous init container recommendation is that each container may need its own distinct TLS components based on the network identity it has. Enabling volume mounts could make sense, but we need to identify what strategies make the most sense to bake into the current init container.

**Comment by aryklein** [comment]
Do you have any plans to support extra configuration or override by specifying some keys on the `PodTemplateSpec`? I'm looking for a way to add additional environment variables to the `cassandra` container.

**Comment by johntrimble** [comment]
For the general case of being able to customize containers such as the cassandra container, I think we could do something similar to what the elastic operator does, in that if we find a container in the user provided podtemplate with the same name as one the operator wants to create, we use the user provided version as a base and apply changes on top of it. I've suggested this before on another ticket:

https://github.com/datastax/cass-operator/issues/183#issuecomment-670074725

**Comment by jimdickinson** [comment]
Fixed by #263 - will be in cass-operator 1.5
